src/wallet/mod.rs

```diff
@@ -253,32 +261,91 @@ where
     }
 
-    // Return the the last previously derived address for `keychain` if it has not been used in a
-    // received transaction. Otherwise return a new address using [`Wallet::get_new_address`].
-    fn get_unused_address(&self, keychain: KeychainKind) -> Result<AddressInfo, Error> {
+    // Return vector of unused address indexes
+    fn get_unused_key_indexes(&self, keychain: KeychainKind) -> Result<Vec<u32>, Error> {
         let current_index = self.fetch_index(keychain)?;
 
-        let derived_key = self
-            .get_descriptor_for_keychain(keychain)
-            .as_derived(current_index, &self.secp);
-
-        let script_pubkey = derived_key.script_pubkey();
+        let derived_keys: Vec<Descriptor<DerivedDescriptorKey>> = (0..current_index + 1)
+            .map(|i| {
+                self.get_descriptor_for_keychain(keychain)
+                    .as_derived(i, &self.secp)
+            })
+            .collect();
 
-        let found_used = self
-            .list_transactions(true)?
+        let unused_key_indexes: Vec<u32> = derived_keys
             .iter()
-            .flat_map(|tx_details| tx_details.transaction.as_ref())
-            .flat_map(|tx| tx.output.iter())
-            .any(|o| o.script_pubkey == script_pubkey);
+            .cloned()
+            .enumerate()
+            .filter_map(|(i, key)| {
+                if !self
+                    .list_transactions(true)
+                    .expect("getting transaction list")
+                    .iter()
+                    .flat_map(|tx_details| tx_details.transaction.as_ref())
+                    .flat_map(|tx| tx.output.iter())
+                    .any(|o| o.script_pubkey == key.script_pubkey())
+                {
+                    Some(i as u32)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        Ok(unused_key_indexes)
+    }
 
-        if found_used {
+    // Return the the last previously derived address for `keychain` if it has not been used in a
+    // received transaction. Otherwise return a new address using [`Wallet::get_new_address`].
+    fn get_last_unused_address(&self, keychain: KeychainKind) -> Result<AddressInfo, Error> {
+        let mut unused_key_indexes = self.get_unused_key_indexes(keychain)?;
+        match unused_key_indexes.pop() {
+            None => self.get_new_address(keychain),
+            Some(address_index) => {
+                let derived_key = self
+                    .get_descriptor_for_keychain(keychain)
+                    .as_derived(address_index, &self.secp);
+
+                let script_pubkey = derived_key.script_pubkey();
+
+                let found_used = self
+                    .list_transactions(true)?
+                    .iter()
+                    .flat_map(|tx_details| tx_details.transaction.as_ref())
+                    .flat_map(|tx| tx.output.iter())
+                    .any(|o| o.script_pubkey == script_pubkey);
```

Comment:
This check is not needed in this function anymore. We already checked that this key wasn't used in `get_unused_key_indexes`..